p2p/net/swarm/connectedness_event_emitter.go

```diff
@@ -0,0 +1,143 @@
+package swarm
+
+import (
+	"context"
+	"sync"
+
+	"github.com/libp2p/go-libp2p/core/event"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// connectednessEventEmitter emits PeerConnectednessChanged events.
+// We ensure that for any peer we connected to we always sent atleast 1 NotConnected Event after
+// the peer disconnects. This is because peers can observe a connection before they are notified
+// of the connection by a peer connectedness changed event.
+type connectednessEventEmitter struct {
+	mx sync.RWMutex
+	// newConns is the channel that holds the peerIDs we recently connected to
+	newConns      chan peer.ID
+	removeConnsMx sync.Mutex
+	// removeConns is a slice of peerIDs we have recently closed connections to
+	removeConns []peer.ID
+	// lastEvent is the last connectedness event sent for a particular peer.
+	lastEvent map[peer.ID]network.Connectedness
+	// connectedness is the function that gives the peers current connectedness state
+	connectedness func(peer.ID) network.Connectedness
+	// emitter is the PeerConnectednessChanged event emitter
+	emitter         event.Emitter
+	wg              sync.WaitGroup
+	removeConnNotif chan struct{}
+	ctx             context.Context
+	cancel          context.CancelFunc
+}
+
+func newConnectednessEventEmitter(connectedness func(peer.ID) network.Connectedness, emitter event.Emitter) *connectednessEventEmitter {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &connectednessEventEmitter{
+		newConns:        make(chan peer.ID, 32),
+		lastEvent:       make(map[peer.ID]network.Connectedness),
+		removeConnNotif: make(chan struct{}, 1),
+		connectedness:   connectedness,
+		emitter:         emitter,
+		ctx:             ctx,
+		cancel:          cancel,
+	}
+	c.wg.Add(1)
+	go c.runEmitter()
+	return c
+}
+
+func (c *connectednessEventEmitter) AddConn(p peer.ID) {
+	c.mx.RLock()
+	defer c.mx.RUnlock()
+	if c.ctx.Err() != nil {
+		return
+	}
+
+	c.newConns <- p
```

Comment:
Did you mean to take a ctx here and exit if ctx is cancelled?

Reply:
I tried, but this turned out very difficult. 

There isn't a nice context to use here. The dial worker loop context doesn't work too well because it expires after a minute(dial timeout) and only blocking for a minute seemed too low. 
Say we have a nice enough context: 
1. we skip sending the connectedness event on expiry, we will still would use the swarm call back notification system to send notifications. I wanted to maintain parity, if we send a callback we will send an event. 
2. we skip the notification callback, this can be done with some extra state on the connection. This extra state is needed because at this point we've added the connection to the conns map. Swarm close at this point will trigger the disconnection callback unless we store the state that we never send the connected callback for this connection. 

We can do 2 and should do 2, by timing out after a few minutes of waiting. I decided not to fix that in this PR because this is the current behavior and this doesn't cause any deadlock since the dial worker loop(which calls this method) is independent of the `DialPeer` call so consumers making dials will only wait for a minute and then will continue consuming and not deadlock.

Reply:
got it. That makes sense 👍